core/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -252,12 +528,24 @@ public void visit(Visitor visitor)
     this.expr = expr;
   }
 
+  abstract UnaryExpr copy(Expr expr);
+
   @Override
   public void visit(Visitor visitor)
   {
     expr.visit(visitor);
     visitor.visit(this);
   }
+
+  @Override
+  public Expr visit(Shuttle shuttle)
+  {
+    Expr newExpr = expr.visit(shuttle);
+    if (newExpr != expr) {
+      return shuttle.visit(copy(newExpr));
```

Comment:
could we document shuttle.visit(..) method to never mutate input arguments and always create new/copy if they wanted to return something different? In that case we wouldn't need the defensive copying and it will happen in shuttle.visit(..) impl only if necessary.
also, for further guarantee, we could mark all fields in all Expr classes to be final , so that they are all Immutable ( I can't recall if there is any need for any of those to be mutable ) . Also document in javadoc of Expr that all derived classed must be immutable.

Reply:
>In that case we wouldn't need the defensive copying and it will happen in shuttle.visit(..) impl only if necessary.

This should already be true, all `Expr` are immutable or are at least intended to be, it was a mistake that some fields were not marked as such, will fix. The copy here is because `Expr` are immutable. For `UnaryExpr`, the `Shuttle` visits the child `Expr` first, and potentially rewrites it to a new `Expr` object. If the returned objects are new objects, we have to create a new `UnaryExpr` with the updated child `Expr`, before allowing the `Shuttle` to visit the `UnaryExpr`. This lets the `Shuttle` accumulate the new `Expr` from the bottom up as it is visiting the `Expr` tree. I can't really think of another way to do it, because there is no guarantee that `shuttle.visit(this)` returns a `UnaryExpr`, it could be anything.

>Also document in javadoc of Expr that all derived classed must be immutable.

:+1: am going to overhaul the javadoc here in general, will be sure to clarify this.

Reply:
I think I understand now why call to copy is necessary . it will probably be more readable if instead of having a copy(..) method , each impls just implemented `visit(Shuttle visit)` method and the common code was refactored into a static utility method in the top level class ... e.g. `UnaryExpr` would contain the utility method and `UnaryMinusExpr` and `UnaryPlusExpr` would just implement `visit(shuttle)` themselves .
That said, these kind of nuances are heavily biased so I leave them to the author.